python/ray/serve/_private/result_wrapper.py

```diff
@@ -0,0 +1,126 @@
+import threading
+import time
+from abc import ABC, abstractmethod
+from typing import Callable, Optional
+
+import ray
+from ray.serve._private.utils import calculate_remaining_timeout
+
+
+class ResultWrapper(ABC):
+    @abstractmethod
+    def get(self, timeout_s: Optional[float]):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def get_async(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def next(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def anext(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def resolve_generator_to_ref_sync(self, timeout_s: Optional[float]):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def resolve_generator_to_ref_async(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def resolve_as_top_level_arg(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def add_callback(self, callback: Callable):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def cancel(self):
+        raise NotImplementedError
+
+
+class ActorResultWrapper(ResultWrapper):
+    def __init__(self, obj_ref_or_gen, is_gen: bool):
+        self._obj_ref = None
+        self._obj_ref_gen = None
+        self._is_gen = is_gen
```

Comment:
Should this just be inferred from the `obj_ref_or_gen`

Reply:
`is_gen` is more like `is_streaming`, which isn't determined by whether obj_ref_or_gen is a generator or not because we unified unary requests onto the streaming path too.

would it be clearer if `is_gen` is instead named `is_streaming`?

Reply:
Yes please

Reply:
renamed!